cdc/sorter/leveldb/table_sorter.go

```diff
@@ -0,0 +1,514 @@
+		tk.Events[deleteKeys[i]] = []byte{}
+	}
+
+	tk.NeedIter = needIter
+	tk.Irange = &lutil.Range{
+		Start: encoding.EncodeTsKey(ls.uid, ls.tableID, 0),
+		Limit: encoding.EncodeTsKey(ls.uid, ls.tableID, maxResolvedTs+1),
+	}
+
+	// Send write task to leveldb.
+	err := ls.router.SendB(ctx, ls.actorID, actormsg.SorterMessage(tk))
+	return tk.IterCh, errors.Trace(err)
+}
+
+// output nonblocking outputs an event. Caller should retry when it returns false.
+func (ls *Sorter) output(event *model.PolymorphicEvent) bool {
+	if ls.lastEvent == nil {
+		ls.lastEvent = event
+	}
+	if ls.lastEvent.CRTs > event.CRTs {
+		log.Panic("regression",
+			zap.Any("lastEntry", ls.lastEvent), zap.Any("event", event),
+			zap.Uint64("regionID", event.RegionID()))
+	}
+	select {
+	case ls.outputCh <- event:
+		ls.lastEvent = event
+		return true
+	default:
+		return false
+	}
+}
+
+// outputResolvedTs nonblocking outputs a resolved ts event.
+func (ls *Sorter) outputResolvedTs(rts model.Ts) {
+	ok := ls.output(model.NewResolvedPolymorphicEvent(0, rts))
+	if ok {
+		ls.lastSentResolvedTs = rts
+	}
+}
+
+// outputBufferedResolvedEvents nonblocking output resolved events and
+// resolved ts that are buffered in outputBuffer.
+// It pops outputted events in the buffer and append their key to deleteKeys.
+func (ls *Sorter) outputBufferedResolvedEvents(
+	buffer *outputBuffer, sendResolvedTsHint bool,
+) {
+	hasRemainEvents := false
+	// Index of remaining output events
+	remainIdx := 0
+	// Commit ts of the last outputed events.
+	lastCommitTs := uint64(0)
+	for idx := range buffer.resolvedEvents {
+		event := buffer.resolvedEvents[idx]
+		ok := ls.output(event)
+		if !ok {
+			hasRemainEvents = true
+			break
+		}
+		lastCommitTs = event.CRTs
+
+		// Delete sent events.
+		key := encoding.EncodeKey(ls.uid, ls.tableID, event)
+		buffer.appendDeleteKey(message.Key(key))
+		remainIdx = idx + 1
+	}
+	// Remove outputted events.
+	buffer.shiftResolvedEvents(remainIdx)
+
+	// If all buffered resolved events are sent, send its resolved ts too.
+	if sendResolvedTsHint && lastCommitTs != 0 && !hasRemainEvents {
+		ls.outputResolvedTs(lastCommitTs)
+	}
+}
+
+// outputBufferedResolvedEvents nonblocking output resolved events and
+// resolved ts that are buffered in leveldb.
+// It appends outputted events's key to outputBuffer deleteKeys to delete them
+// later, and appends not-yet-send resolved events to outputBuffer resolvedEvents
+// to send them later.
+// outputBuffer must be empty.
+func (ls *Sorter) outputIterEvents(
+	iter message.LimitedIterator, buffer *outputBuffer, maxResolvedTs uint64,
+) uint64 {
+	lenResolvedEvents, lenDeleteKeys := buffer.len()
+	if lenDeleteKeys > 0 || lenResolvedEvents > 0 {
+		log.Panic("buffer is not empty",
+			zap.Int("deleteKeys", lenDeleteKeys),
+			zap.Int("resolvedEvents", lenResolvedEvents))
+	}
+
+	// The commit ts of buffered resolved events.
+	lastCommitTs := uint64(0)
+	iterHasNext := iter.Next()
+SEEK_SEND:
```

Comment:
It seems like we should use camel case here.

See: https://golang.org/doc/effective_go#switch